modules/microshift-oc-mirror-embed-ops-offline-use.adoc

```diff
@@ -0,0 +1,159 @@
+//Module included in the following assemblies:
+//
+//* microshift_running_apps/microshift-operators-olm.adoc
+
+:_mod-docs-content-type: PROCEDURE
+[id="microshift-oc-mirror-embed-ops-offline-use_{context}"]
+= Using catalog and Operator images offline in RHEL for Edge
+
+To use widely available Operators in a {op-system-ostree-first}, use oc-mirror in dry-run mode to produce the filtered down catalog index. After the dry run, you must parse the index, get all the image references, and format the output for adding to an Image Builder blueprint.
+
+[NOTE]
+====
+For custom-made catalogs and Operators, you can format image references for the Image Builder blueprint without using the following procedure.
+====
+
+.Prerequisites
+* You have a catalog index for the Operators you want to use.
+* jq is installed.
+* You have an Image Builder blueprint TOML file. If you have not created one, you can use the blueprint installed in the `/usr/share/microshift/blueprint` directory that is specific to your platform architecture.
+
+.Procedure
+
+. Parse the catalog index and get the image references by running the following example command:
++
+[source,terminal]
+----
+jq -r --slurp '.[] | select(.relatedImages != null) | "[[containers]]\nsource = \"" + .relatedImages[].image + "\"\n"'   ./oc-mirror-workspace/src/catalogs/registry.redhat.io/redhat/redhat-operator-index/v4.14/index/index.json
+----
+
+. Optional: Filter out images that cannot be mirrored by running the following example command:
++
+[source,terminal]
+----
+$ jq -r --slurp '.[] | select(.relatedImages != null) | .relatedImages[] | select(.name |  contains("ppc") or contains("s390x") | not) | "[[containers]]\\nsource = \\"" + .image + "\\"\\n"' ./oc-mirror-workspace/src/catalogs/registry.redhat.io/redhat/redhat-operator-index/v4.14/index/index.json
+----
++
+[NOTE]
+====
+This optional step uses the AMQ Broker Operator as an example. You can add other criteria to the `jq` command for further filtering as required by your use case.
```

Comment:
@ShaunaDiaz 
Just FYI.
I am not sure we should add more to reminder the user to double check if all the sources got from index.json are digest. if not, the installation of operator will fail.

@pmtk how about your opinion?

Reply:
From the oc-mirror perspective, files under oc-mirror-workspace/src shouldn't really be used:
* They are kept in the folder for debugging reasons only.
*  They are internal to oc-mirror logic.
* They are destroyed normally at the end of the execution
* They will not be generated anymore in the v2
* You can get similar output from 
```bash
opm render registry.redhat.io/redhat/redhat-operator-index:v4.14
```

Reply:
@kuiwang02 that's a good point. For mirrored/offline use cases digests are strongly preferred - with tags something could always want to check with upstream registry and fail.

@sherine-k Unfortunately `opm render` just mirrors the index, it doesn't filter it so it's not useful here...

Reply:
Yes, and opm is deprecated, isn't it? https://docs.openshift.com/container-platform/4.15/operators/understanding/olm-rh-catalogs.html MicroShift users are creating file-based catalogs in connected circumstances, so I don't think we want to flip the boat for disconnected. Would be very confusing for customers.

Reply:
@ShaunaDiaz opm is not deprecated. the deprecated is sqllite-based format.

@pmtk here is what I try with opm to get image lists. FYI
1, render json from index image
```console
[root@preserve-olm-env2 bin]# opm render registry.redhat.io/redhat/redhat-operator-index:v4.15 -o json> 415.json
```
2, get bundle name list per package and channel. take amq-broker-rhel8 and 7.11.x as example
```console
[root@preserve-olm-env2 bin]# jq -s '.[] | select( .package == "amq-broker-rhel8" ) | select( .schema == "olm.channel" ) | select( .name == "7.11.x" ) | .entries | .[] | .name' ./415.json
"amq-broker-operator.v7.11.0-opr-2"
"amq-broker-operator.v7.11.0-opr-3"
"amq-broker-operator.v7.11.1-opr-1"
"amq-broker-operator.v7.11.1-opr-2"
"amq-broker-operator.v7.11.2-opr-1"
"amq-broker-operator.v7.11.3-opr-1"
"amq-broker-operator.v7.11.4-opr-1"
"amq-broker-operator.v7.11.5-opr-1"
"amq-broker-operator.v7.11.6-opr-2"
```

3, get image list per bundle. take amq-broker-operator.v7.11.0-opr-2 as example
```console
[root@preserve-olm-env2 bin]# jq -s '.[] | select( .schema == "olm.bundle" ) | select ( .name == "amq-broker-operator.v7.11.0-opr-2") | select( .package == "amq-broker-rhel8") | .relatedImages[] | .image' ./415.json
"registry.redhat.io/amq7/amq-broker-init-openj9-11-rhel8@sha256:1a9f13fd565113055b2aa2022bd0709a63c1dc058a8570e0fd7eba5c7b101334"
"registry.redhat.io/amq7/amq-broker-init-openj9-11-rhel8@sha256:1a9f13fd565113055b2aa2022bd0709a63c1dc058a8570e0fd7eba5c7b101334"
"registry.redhat.io/amq7/amq-broker-init-openj9-11-rhel8@sha256:2cb14c917246ffcbd5d861ca6f3d58cb1146a4e55715ee08160e115d3dfc1c3e"
"registry.redhat.io/amq7/amq-broker-init-openj9-11-rhel8@sha256:2cb14c917246ffcbd5d861ca6f3d58cb1146a4e55715ee08160e115d3dfc1c3e"
"registry.redhat.io/amq7/amq-broker-init-openj9-11-rhel8@sha256:56282c6a461e945a83bffb13d3f3d8d9621800f0fc9cd755a75d21eacd8f5fbf"
"registry.redhat.io/amq7/amq-broker-init-openj9-11-rhel8@sha256:56282c6a461e945a83bffb13d3f3d8d9621800f0fc9cd755a75d21eacd8f5fbf"
"registry.redhat.io/amq7/amq-broker-init-openj9-11-rhel8@sha256:5ff28e3062d3eabb9f52d8271b837fdab06cafbd8b8f732b1bc00de49ad69aad"
"registry.redhat.io/amq7/amq-broker-init-openj9-11-rhel8@sha256:5ff28e3062d3eabb9f52d8271b837fdab06cafbd8b8f732b1bc00de49ad69aad"
"registry.redhat.io/amq7/amq-broker-init-openj9-11-rhel8@sha256:9465e398a55c89c121b3ae50374414190a991839fa5e1018fc02551f6aa3b9aa"
"registry.redhat.io/amq7/amq-broker-init-openj9-11-rhel8@sha256:9465e398a55c89c121b3ae50374414190a991839fa5e1018fc02551f6aa3b9aa"
"registry.redhat.io/amq7/amq-broker-init-openj9-11-rhel8@sha256:c3ac4602b48867f3d4f7cc0efd5c9f133098e409626cfb402f80386ff460dbf4"
"registry.redhat.io/amq7/amq-broker-init-openj9-11-rhel8@sha256:c3ac4602b48867f3d4f7cc0efd5c9f133098e409626cfb402f80386ff460dbf4"
"registry.redhat.io/amq7/amq-broker-init-openjdk-11-rhel8@sha256:aa7af7e038a12b5202108b362c8af243954603e84034a16c0192684c86d5d28b"
"registry.redhat.io/amq7/amq-broker-init-openjdk-11-rhel8@sha256:aa7af7e038a12b5202108b362c8af243954603e84034a16c0192684c86d5d28b"
"registry.redhat.io/amq7/amq-broker-init-openjdk-11-rhel8@sha256:d4a9176f49ffd9db6051fbda15d47a1aa32c282fb5010d0caf5b9243e80bf055"
"registry.redhat.io/amq7/amq-broker-init-openjdk-11-rhel8@sha256:d4a9176f49ffd9db6051fbda15d47a1aa32c282fb5010d0caf5b9243e80bf055"
"registry.redhat.io/amq7/amq-broker-init-rhel8@sha256:0b2126cfb6054fdf428c1f43b69e36e93a09a49ce15350e9273c98cc08c6598b"
"registry.redhat.io/amq7/amq-broker-init-rhel8@sha256:34739366153b825dcb113e0ed6115ef5656085cf8d47492789ff17464042b576"
"registry.redhat.io/amq7/amq-broker-init-rhel8@sha256:34739366153b825dcb113e0ed6115ef5656085cf8d47492789ff17464042b576"
"registry.redhat.io/amq7/amq-broker-init-rhel8@sha256:34739366153b825dcb113e0ed6115ef5656085cf8d47492789ff17464042b576"
"registry.redhat.io/amq7/amq-broker-init-rhel8@sha256:45a5e054546a2c1ee338518f8477f728f5f7108c9dd8b516b4c79d4a7d6f5ba7"
"registry.redhat.io/amq7/amq-broker-init-rhel8@sha256:4e09ee3bd8ff3479fdd00e221b7b1505ae04ca6c64fd6bc9e2f1a34b8e98d137"
"registry.redhat.io/amq7/amq-broker-init-rhel8@sha256:61e5a8c1e46f1c18648e0406e2a57c52f911979079020cc57c7a4b06a56940bb"
"registry.redhat.io/amq7/amq-broker-init-rhel8@sha256:869cda365447b0b7b53ed21d2496884c15dc9c07295a9dfa0337c2c9974fe018"
"registry.redhat.io/amq7/amq-broker-init-rhel8@sha256:869cda365447b0b7b53ed21d2496884c15dc9c07295a9dfa0337c2c9974fe018"
"registry.redhat.io/amq7/amq-broker-init-rhel8@sha256:869cda365447b0b7b53ed21d2496884c15dc9c07295a9dfa0337c2c9974fe018"
"registry.redhat.io/amq7/amq-broker-init-rhel8@sha256:b575d6eebf44dc9049445e357488efb17a690953852419b534df30e0105b9ca4"
"registry.redhat.io/amq7/amq-broker-init-rhel8@sha256:b575d6eebf44dc9049445e357488efb17a690953852419b534df30e0105b9ca4"
"registry.redhat.io/amq7/amq-broker-init-rhel8@sha256:b575d6eebf44dc9049445e357488efb17a690953852419b534df30e0105b9ca4"
"registry.redhat.io/amq7/amq-broker-init-rhel8@sha256:b74d03ed852a3731467ffda95266ce49f2065972f1c37bf254f3d52b34c11991"
"registry.redhat.io/amq7/amq-broker-init-rhel8@sha256:ba21779c29b431a22c824e91820705c9501fb0b7d6f7e1ae22fcd3becc8c8a2f"
"registry.redhat.io/amq7/amq-broker-init-rhel8@sha256:c867ec9ac1199431a10b251699b30a7d60567f2df48a9825002c8216281775a7"
"registry.redhat.io/amq7/amq-broker-init-rhel8@sha256:c867ec9ac1199431a10b251699b30a7d60567f2df48a9825002c8216281775a7"
"registry.redhat.io/amq7/amq-broker-init-rhel8@sha256:c867ec9ac1199431a10b251699b30a7d60567f2df48a9825002c8216281775a7"
"registry.redhat.io/amq7/amq-broker-init-rhel8@sha256:e03945756d86bd189e8c33027fa668d2065c168510085dfd06defd9f40d3e0ab"
"registry.redhat.io/amq7/amq-broker-init-rhel8@sha256:f58678f03a250dbb296ebeb112c9150340155799ba4dad9e90bf347b9ed0ca7b"
"registry.redhat.io/amq7/amq-broker-openj9-11-rhel8@sha256:34ab1f66d8daa544e3823464383e9788e307665c875499838fbed95f3a74a4dd"
"registry.redhat.io/amq7/amq-broker-openj9-11-rhel8@sha256:34ab1f66d8daa544e3823464383e9788e307665c875499838fbed95f3a74a4dd"
"registry.redhat.io/amq7/amq-broker-openj9-11-rhel8@sha256:4440b20349032bb5c0bb2a8deb89832b37767af6f0b3208b067867a85641913e"
"registry.redhat.io/amq7/amq-broker-openj9-11-rhel8@sha256:4440b20349032bb5c0bb2a8deb89832b37767af6f0b3208b067867a85641913e"
"registry.redhat.io/amq7/amq-broker-openj9-11-rhel8@sha256:97c3aa5673610875a53571449f7d5210ae8d675072d044f17b416725ed0dba88"
"registry.redhat.io/amq7/amq-broker-openj9-11-rhel8@sha256:97c3aa5673610875a53571449f7d5210ae8d675072d044f17b416725ed0dba88"
"registry.redhat.io/amq7/amq-broker-openj9-11-rhel8@sha256:ba0f5df7da08bc030ff5e0e68ecb52a931c25f6b67c6556aa8a9661c82d69d42"
"registry.redhat.io/amq7/amq-broker-openj9-11-rhel8@sha256:ba0f5df7da08bc030ff5e0e68ecb52a931c25f6b67c6556aa8a9661c82d69d42"
"registry.redhat.io/amq7/amq-broker-openj9-11-rhel8@sha256:d4010af53429502f41d722c39d69a7bb147bb06d19e3d4a69de70dbe88e3990e"
"registry.redhat.io/amq7/amq-broker-openj9-11-rhel8@sha256:d4010af53429502f41d722c39d69a7bb147bb06d19e3d4a69de70dbe88e3990e"
"registry.redhat.io/amq7/amq-broker-openj9-11-rhel8@sha256:f2651cb097eb631fc6af90e4a8404d36516aa2205c546125bf0c37739ab525fa"
"registry.redhat.io/amq7/amq-broker-openj9-11-rhel8@sha256:f2651cb097eb631fc6af90e4a8404d36516aa2205c546125bf0c37739ab525fa"
"registry.redhat.io/amq7/amq-broker-openjdk-11-rhel8@sha256:7abfdcfd7b2d3d12b9234f7b057bd199bceefa47bdf02f62236f82a13fdbf5b7"
"registry.redhat.io/amq7/amq-broker-openjdk-11-rhel8@sha256:7abfdcfd7b2d3d12b9234f7b057bd199bceefa47bdf02f62236f82a13fdbf5b7"
"registry.redhat.io/amq7/amq-broker-openjdk-11-rhel8@sha256:967d840b831b310abd3a468a8652d1fa48b2ce28b342bfab53999245deb004f1"
"registry.redhat.io/amq7/amq-broker-openjdk-11-rhel8@sha256:967d840b831b310abd3a468a8652d1fa48b2ce28b342bfab53999245deb004f1"
"registry.redhat.io/amq7/amq-broker-rhel8-operator-bundle@sha256:038489a08c1d982197c47c58b2fe7ea355029ef5aab0fedd3050cec0ba1540ce"
"registry.redhat.io/amq7/amq-broker-rhel8-operator@sha256:30c08e16131aa4734c3da44ddf687aea15b01e31c42cb32f818fd845c4d5f7f4"
"registry.redhat.io/amq7/amq-broker-rhel8@sha256:0d857b95084e39c011651be89482d4dcc77c2d1c3ac3c6a819f3c9d5a57a425f"
"registry.redhat.io/amq7/amq-broker-rhel8@sha256:3219f90789690a3c5cfcc86e5f602c443e75f1fdb3a19660f43ee3996bfa8f62"
"registry.redhat.io/amq7/amq-broker-rhel8@sha256:3c00b524722b500ff0922fe7a8722868557512183a8f69248f37a68c45c6a770"
"registry.redhat.io/amq7/amq-broker-rhel8@sha256:3c00b524722b500ff0922fe7a8722868557512183a8f69248f37a68c45c6a770"
"registry.redhat.io/amq7/amq-broker-rhel8@sha256:3c00b524722b500ff0922fe7a8722868557512183a8f69248f37a68c45c6a770"
"registry.redhat.io/amq7/amq-broker-rhel8@sha256:5d39f26f06769fe5c989ea0e0ee8f9166120e152ce2c9bc3336fc8d7cab2ff92"
"registry.redhat.io/amq7/amq-broker-rhel8@sha256:67ae74e72927557c55611d0248347b2099b117ac1aa245dac05856fe782fb59c"
"registry.redhat.io/amq7/amq-broker-rhel8@sha256:6e9e0e3fddc44fbf87591d8904176407f1ec409d578a7d108c13cdd48d85131b"
"registry.redhat.io/amq7/amq-broker-rhel8@sha256:86ff145bd68e797a0507495570063b54cbe4186fcb4acd83c900406619ef9524"
"registry.redhat.io/amq7/amq-broker-rhel8@sha256:86ff145bd68e797a0507495570063b54cbe4186fcb4acd83c900406619ef9524"
"registry.redhat.io/amq7/amq-broker-rhel8@sha256:86ff145bd68e797a0507495570063b54cbe4186fcb4acd83c900406619ef9524"
"registry.redhat.io/amq7/amq-broker-rhel8@sha256:979b59325aa0f34eb05625201beba53fccbb83bd5eb80a89dcb5261ae358138f"
"registry.redhat.io/amq7/amq-broker-rhel8@sha256:ae79b5dab78c4f48993fd8999b09bdcaa918b8e30a9f750bf24d411710f71928"
"registry.redhat.io/amq7/amq-broker-rhel8@sha256:c5f4c08e068b9721967cf7c7cbd9a9e93fb5e39b264dd13b653e99d8f3fa9e0e"
"registry.redhat.io/amq7/amq-broker-rhel8@sha256:c9eba0d1f9fba588711774f4db75221f057578407065127ed60c33eef564f499"
"registry.redhat.io/amq7/amq-broker-rhel8@sha256:c9eba0d1f9fba588711774f4db75221f057578407065127ed60c33eef564f499"
"registry.redhat.io/amq7/amq-broker-rhel8@sha256:c9eba0d1f9fba588711774f4db75221f057578407065127ed60c33eef564f499"
"registry.redhat.io/amq7/amq-broker-rhel8@sha256:f970887546fdad5ab8eff991d365b3e15327fa1a592b663a72dbe60f2f7fbee3"
"registry.redhat.io/amq7/amq-broker-rhel8@sha256:f970887546fdad5ab8eff991d365b3e15327fa1a592b663a72dbe60f2f7fbee3"
"registry.redhat.io/amq7/amq-broker-rhel8@sha256:f970887546fdad5ab8eff991d365b3e15327fa1a592b663a72dbe60f2f7fbee3"
[root@preserve-olm-env2 bin]# 

```

with the above, we could get image list.
maybe need oneline shell script to do it in one command.

Reply:
here is the example for one line
```console
[root@preserve-olm-env2 bin]# images=();for bundleName in $(jq -r -s '.[] | select( .package == "amq-broker-rhel8" ) | select( .schema == "olm.channel" ) | select( .name == "7.11.x" ) | .entries | .[] | .name' ./415.json); do images+=( $(jq -r --arg bundleName "$bundleName" -s '.[] | select( .schema == "olm.bundle" ) | select ( .name == $bundleName) | select( .package == "amq-broker-rhel8") | .relatedImages[] | .image' ./415.json) ); done;unique_images=( $(printf "%s\n" "${images[@]}" | sort -u) );printf "%s\n" "${unique_images[@]}"
registry.redhat.io/amq7/amq-broker-init-openj9-11-rhel8@sha256:1a9f13fd565113055b2aa2022bd0709a63c1dc058a8570e0fd7eba5c7b101334
registry.redhat.io/amq7/amq-broker-init-openj9-11-rhel8@sha256:2cb14c917246ffcbd5d861ca6f3d58cb1146a4e55715ee08160e115d3dfc1c3e
registry.redhat.io/amq7/amq-broker-init-openj9-11-rhel8@sha256:56282c6a461e945a83bffb13d3f3d8d9621800f0fc9cd755a75d21eacd8f5fbf
registry.redhat.io/amq7/amq-broker-init-openj9-11-rhel8@sha256:5ff28e3062d3eabb9f52d8271b837fdab06cafbd8b8f732b1bc00de49ad69aad
registry.redhat.io/amq7/amq-broker-init-openj9-11-rhel8@sha256:9465e398a55c89c121b3ae50374414190a991839fa5e1018fc02551f6aa3b9aa
registry.redhat.io/amq7/amq-broker-init-openj9-11-rhel8@sha256:c3ac4602b48867f3d4f7cc0efd5c9f133098e409626cfb402f80386ff460dbf4
registry.redhat.io/amq7/amq-broker-init-openjdk-11-rhel8@sha256:aa7af7e038a12b5202108b362c8af243954603e84034a16c0192684c86d5d28b
registry.redhat.io/amq7/amq-broker-init-openjdk-11-rhel8@sha256:d4a9176f49ffd9db6051fbda15d47a1aa32c282fb5010d0caf5b9243e80bf055
registry.redhat.io/amq7/amq-broker-init-rhel8@sha256:0b2126cfb6054fdf428c1f43b69e36e93a09a49ce15350e9273c98cc08c6598b
registry.redhat.io/amq7/amq-broker-init-rhel8@sha256:0dde839c2dce7cb684094bf26523c8e16677de03149a0fff468b8c3f106e1f4f
...
registry.redhat.io/amq7/amq-broker-rhel8@sha256:f970887546fdad5ab8eff991d365b3e15327fa1a592b663a72dbe60f2f7fbee3
registry.redhat.io/amq7/amq-broker-rhel8@sha256:ff6fefad518a6c997d4c5a6e475ba89640260167f0bc27715daf3cc30116fad1
[root@preserve-olm-env2 bin]#
```

Reply:
@kuiwang02 thanks for providing alternative method!

It's nice this doesn't use internal files, but now I'm a bit puzzled how should we present this to customer. There are several inputs in this invocation and they are very dependent of each other: package > channel > version > images (and not all of them can be embedded due to ARCH mismatch). I fell like we're missing some real tool (that has capabilities of v1 oc-mirror plus more) so it's nicely packaged rather than bunch or jq or yq executions...

Reply:
@pmtk yes. it is complicated to use jq to get required information for customer.
I just find another way. FYI.
although oc-mirror does not generate src, it still generate mapping.txt when we run oc-mirror with dry-run (Per my understanding, it must generate it), and then we could get images list.
here is what I did (FYI in the future):
```console
[root@preserve-olm-env2 bin]# cat isc.yaml 
kind: ImageSetConfiguration
apiVersion: mirror.openshift.io/v1alpha2
storageConfig:
  local:
    path: /tmp/micorshiftolmindexmirror
mirror:
  operators:
  - catalog: registry.redhat.io/redhat/redhat-operator-index:v4.15 
    packages:
    - name: amq-broker-rhel8 
      channels:
      - name: 7.11.x 
[root@preserve-olm-env2 bin]# oc mirror --config=./isc.yaml docker://quay.io/olmqe --dry-run
Checking push permissions for quay.io
Creating directory: oc-mirror-workspace/src/publish
Creating directory: oc-mirror-workspace/src/v2
Creating directory: oc-mirror-workspace/src/charts
Creating directory: oc-mirror-workspace/src/release-signatures
No metadata detected, creating new workspace
wrote mirroring manifests to oc-mirror-workspace/operators.1711506376/manifests-redhat-operator-index

To upload local images to a registry, run:

	oc adm catalog mirror file://redhat/redhat-operator-index:v4.15 REGISTRY/REPOSITORY
quay.io/
  olmqe/amq7/amq-broker-init-openj9-11-rhel8
    blobs:
      registry.redhat.io/amq7/amq-broker-init-openj9-11-rhel8 sha256:53ae128587a66ae4816f2dc2df684d892b1a79f0dda7e5d49ed1fcb445857926 1.742KiB
      registry.redhat.io/amq7/amq-broker-init-openj9-11-rhel8 sha256:56ed4da7b1ef50557d54ec71f32a7865963b27789e121c19613b51a01e41e3cb 1.745KiB
...
phase 0:
  quay.io olmqe/amq7/amq-broker-rhel8-operator-bundle blobs=2   mounts=0 manifests=1  shared=0
  quay.io olmqe/amq7/amq-broker-rhel8-operator        blobs=9   mounts=0 manifests=5  shared=0
  quay.io olmqe/amq7/amq-broker-init-rhel8            blobs=270 mounts=0 manifests=89 shared=176
  quay.io olmqe/amq7/amq-broker-init-openj9-11-rhel8  blobs=72  mounts=0 manifests=24 shared=40
  quay.io olmqe/amq7/amq-broker-openjdk-11-rhel8      blobs=20  mounts=0 manifests=8  shared=16
phase 1:
  quay.io olmqe/amq7/amq-broker-rhel8                 blobs=223 mounts=176 manifests=89 shared=176
  quay.io olmqe/amq7/amq-broker-openj9-11-rhel8       blobs=56  mounts=40  manifests=24 shared=40
  quay.io olmqe/amq7/amq-broker-init-openjdk-11-rhel8 blobs=24  mounts=16  manifests=8  shared=16

info: Planning completed in 26.87s
info: Dry run complete
Writing image mapping to oc-mirror-workspace/mapping.txt
[root@preserve-olm-env2 bin]# cd oc-mirror-workspace/
[root@preserve-olm-env2 oc-mirror-workspace]# tree
.
└── mapping.txt

0 directories, 1 file
[root@preserve-olm-env2 oc-mirror-workspace]# cat mapping.txt 
registry.redhat.io/amq7/amq-broker-init-rhel8@sha256:34739366153b825dcb113e0ed6115ef5656085cf8d47492789ff17464042b576=quay.io/olmqe/amq7/amq-broker-init-rhel8:3104d776
registry.redhat.io/amq7/amq-broker-init-rhel8@sha256:ba21779c29b431a22c824e91820705c9501fb0b7d6f7e1ae22fcd3becc8c8a2f=quay.io/olmqe/amq7/amq-broker-init-rhel8:ed80bae4
registry.redhat.io/amq7/amq-broker-rhel8-operator-bundle@sha256:9d0c3253b045f9336a10937c9203eb6d2986cb09e718b6c154530732299bfaa6=quay.io/olmqe/amq7/amq-broker-rhel8-operator-bundle:6534c3b1
registry.redhat.io/amq7/amq-broker-rhel8@sha256:c9eba0d1f9fba588711774f4db75221f057578407065127ed60c33eef564f499=quay.io/olmqe/amq7/amq-broker-rhel8:db3b4de7
registry.redhat.io/amq7/amq-broker-init-openjdk-11-rhel8@sha256:aa7af7e038a12b5202108b362c8af243954603e84034a16c0192684c86d5d28b=quay.io/olmqe/amq7/amq-broker-init-openjdk-11-rhel8:1950ef0b
registry.redhat.io/amq7/amq-broker-rhel8-operator@sha256:6a87f1a46da682870c4495bece5d8ef5b2fb62b4ecda9e6826664031757bf984=quay.io/olmqe/amq7/amq-broker-rhel8-operator:877025c3
registry.redhat.io/amq7/amq-broker-rhel8@sha256:3219f90789690a3c5cfcc86e5f602c443e75f1fdb3a19660f43ee3996bfa8f62=quay.io/olmqe/amq7/amq-broker-rhel8:1db339bf
...
registry.redhat.io/amq7/amq-broker-rhel8@sha256:67ae74e72927557c55611d0248347b2099b117ac1aa245dac05856fe782fb59c=quay.io/olmqe/amq7/amq-broker-rhel8:79587b10
registry.redhat.io/amq7/amq-broker-init-rhel8@sha256:8b4c4400a02c6cacfd978147b158e2ccbe5da955578b4b970a5ee7edac865326
registry.redhat.io/amq7/amq-broker-rhel8@sha256:89bb769b4c33eb931cc7866eeabec9a315b913851cfa84bc6e11ae8d1e8b2cbb
registry.redhat.io/amq7/amq-broker-rhel8@sha256:67ae74e72927557c55611d0248347b2099b117ac1aa245dac05856fe782fb59c[root@preserve-olm-env2 oc-mirror-workspace]# cat mapping.txt | cut -d"=" -f1
registry.redhat.io/amq7/amq-broker-init-rhel8@sha256:34739366153b825dcb113e0ed6115ef5656085cf8d47492789ff17464042b576
registry.redhat.io/amq7/amq-broker-init-rhel8@sha256:ba21779c29b431a22c824e91820705c9501fb0b7d6f7e1ae22fcd3becc8c8a2f
registry.redhat.io/amq7/amq-broker-rhel8-operator-bundle@sha256:9d0c3253b045f9336a10937c9203eb6d2986cb09e718b6c154530732299bfaa6
registry.redhat.io/amq7/amq-broker-rhel8@sha256:c9eba0d1f9fba588711774f4db75221f057578407065127ed60c33eef564f499
```
with this, we do not depend on opm, and still use oc-mirror to get list.

@ShaunaDiaz 
per this decision, i am ok as https://github.com/openshift/openshift-docs/pull/71992#issuecomment-2019290693. note: please do not cherry-pick this PR to 4.14. Thanks

Reply:
Thanks @kuiwang02 for another alternative!

I'm afraid we cannot use it as well.
List of images in `mapping.txt` doesn't have the names and unfortunately for amq-broker it matters because the images are not multi-arch -- they have arch in the name and we cannot feed wrong arch to the blueprints.


```
registry.redhat.io/amq7/amq-broker-rhel8@sha256:c9eba0d1f9fba588711774f4db75221f057578407065127ed60c33eef564f499
```

```
{
  "name": "activemq_artemis_broker_kubernetes_7102_s390x",
  "image": "registry.redhat.io/amq7/amq-broker-rhel8@sha256:c9eba0d1f9fba588711774f4db75221f057578407065127ed60c33eef564f499"
}
```

Reply:
@pmtk thanks for information.
so, if there is more more support from oc-mirror, I have no more idea besides depending on jq and json generated by op.